x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticationToken.java

```diff
@@ -9,145 +9,85 @@
 import com.nimbusds.jwt.JWTClaimsSet;
 import com.nimbusds.jwt.SignedJWT;
 
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
-import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.settings.SecureString;
 import org.elasticsearch.core.Nullable;
 import org.elasticsearch.xpack.core.security.authc.AuthenticationToken;
-import org.elasticsearch.xpack.core.security.authc.jwt.JwtRealmsServiceSettings;
 
 import java.text.ParseException;
-import java.util.List;
-import java.util.Map;
-import java.util.TreeSet;
-import java.util.stream.Collectors;
+import java.util.Arrays;
+import java.util.Objects;
 
 /**
  * An {@link AuthenticationToken} to hold JWT authentication related content.
  */
 public class JwtAuthenticationToken implements AuthenticationToken {
-    private static final Logger LOGGER = LogManager.getLogger(JwtAuthenticationToken.class);
-
-    // Stored members
-    protected SecureString endUserSignedJwt; // required
-    protected SecureString clientAuthenticationSharedSecret; // optional, nullable
-    protected String principal; // Defaults to "iss/aud/sub", with an ordered "aud" list
+    private final String principal;
+    private SignedJWT signedJWT;
+    private final byte[] userCredentialsHash;
+    @Nullable
+    private final SecureString clientAuthenticationSharedSecret;
 
     /**
-     * Store a mandatory JWT and optional Shared Secret. Parse the JWT, and extract the header, claims set, and signature.
-     * Compute a token principal, for use as a realm order cache key. For OIDC ID Tokens, cache key is iss/aud/sub.
-     * For other JWTs, {@link JwtRealmsServiceSettings#PRINCIPAL_CLAIMS_SETTING} supports alternative claims for sub.
-     * Throws IllegalArgumentException if principalClaimNames is empty, JWT is missing, or if JWT parsing fails.
-     * @param principalClaimNames Ordered list of string claims to use for principalClaimValue. The first one found is used (ex: sub).
-     * @param endUserSignedJwt Base64Url-encoded JWT for End-user authentication. Required by all JWT realms.
+     * Store a mandatory JWT and optional Shared Secret.
+     * @param principal The token's principal, useful as a realm order cache key
+     * @param signedJWT The JWT parsed from the end-user credentials
+     * @param userCredentialsHash The hash of the end-user credentials
```

Comment:
This param doc doesn't explain much - I don't really learn anything that the parameter name didn't already tell me.

Reply:
I expanded it to explain it is used for user cache key computation and add a reference to `JwtRealm#authenticate`. Thanks!